runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,642 @@
+    public boolean containsValue(Object value) {
+        return mainMap.containsValue(value);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public V get(@ConflictParameter Object key) {
+        return mainMap.get(key);
+    }
+
+
+
+    /** Get a mapping using the specified index function.
+     *
+     * @param indexFunction The index function to use.
+     * @param index         The index
+     * @return
+     */
+    @SuppressWarnings("unchecked")
+    @DontInstrument
+    public @Nonnull Collection<Object> getByIndex(@Nonnull F indexFunction, I index) {
+        return getByIndex(indexFunction, indexFunction.getProjectionFunction(), index);
+    }
+
+    /** Get a mapping using the specified index function.
+     *
+     * @param indexFunction The index function to use.
+     * @param projectionFunction    The function to project results with.
+     * @param index         The index
+     * @param <P>                   The projection return type.
+     * @return
+     */
+    @DontInstrument
+    public @Nonnull <P> Collection<P> getByIndex(@Nonnull F indexFunction,
+                                @Nonnull ProjectionFunction<K, V, I, P> projectionFunction,
+                                I index) {
+        return getByIndexAndFilter(indexFunction, projectionFunction, e -> true, index);
+    }
+
+    /** Scan and filter using the specified index function and projection.
+     *
+     * @param indexFunction         An indexing function.
+     * @param projectionFunction    The function to project results with.
+     * @param entryPredicate        The predicate to scan and filter with.
+     * @param index                 The index to lookup.
+     * @param <P>                   The projection return type.
+     * @return
+     */
+    @Accessor
+    public @Nonnull <P> Collection<P> getByIndexAndFilter(@Nonnull F indexFunction,
+                                 @Nonnull ProjectionFunction<K, V, I, P> projectionFunction,
+                                 @Nonnull Predicate<? super Map.Entry<K, V>>
+                                                                  entryPredicate,
+                                 I index) {
+        return projectionFunction
+                .generateProjection(index, indexMap
+                        .get(indexFunction).get(index).parallelStream()
+                        .filter(entryPredicate))
+                .collect(Collectors.toList());
+    }
+
+
+    /**
+     * Register new index class
+     *
+     * This replaces the current index.
+     *
+     * @param indexFunctionEnumClass
+     */
+    public void registerIndex(Class indexFunctionEnumClass) {
+        indexerClass = indexFunctionEnumClass;
+        indexMap = new HashMap<>();
+        indexFunctions = EnumSet.allOf(indexFunctionEnumClass);
+        indexFunctions.forEach(f -> indexMap.put(f, ArrayListMultimap.create()));
+        mainMap.forEach(this::mapSecondaryIndexes);
+    }
+
+    /** {@inheritDoc} */
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)